test/topology/test_cluster_features.py

```diff
@@ -145,7 +145,7 @@ async def test_joining_old_node_fails(manager: ManagerClient) -> None:
     await manager.server_start(new_server_info.server_id, expected_error="Feature check failed")
 
     # Try to replace with a node that doesn't support the feature - should fail
-    await manager.server_stop(servers[0].server_id)
+    await manager.server_stop_gracefully(servers[0].server_id)
```

Comment:
I don't like this.

We should be testing cluster behavior under abrupt failures (e.g. `kill -9`), we can't rely on graceful shutdowns to happen each time.

Each time I review this PR I have more and more of this feeling that we're doing something wrong...

First of all, using gossiper liveness information is of course unreliable. What happens when node is marked down by gossiper, not because the node is down, but is behind a network partition? What will the topology coordinator do in this case? I guess that it will attempt to do the replace operation and ban the node in the process. What if the node comes back in the middle of replace? Will it disrupt the cluster while replace is ongoing? Will it send/receive data path operations? Maybe the correct way to handle it is to ban it *at the beginning* of replace operation, not at the end as it is now (if I remember correctly)? @gleb-cloudius WDYT?

I think it's understood that we should be using this stuff only for improving user experience, not relying on it for safety (which is impossible).

So fine, we can:
- reject replace if node is considered alive
- and still have a mechanism that protects us (like early node ban) in case we replace an actually alive (or just temporarily dead) node.

Please check if we need to do that second thing, if it's not done already (do the experiment with having a replaced node come back to life in the middle / at the beginning of replace). If not, please create another follow-up issue.

Second, regarding the abrupt stop. We should have tests with abrupt stop. At least one. Of course waiting 20 seconds because of the user-experience gossiper-based protection is not an option. I see the following options:
- this 20 seconds delay is failure detector setting, which can be configured. We could move these replace tests into `topology_custom` and start clusters there with custom configuration, reducing the delay. This should go together with disabling the tests in debug mode, where the low delay combined with latencies may cause failure detector "flickering"
- OR disable the protection using a failure injection, but that probably also requires moving the tests to `topology_custom`, to enable the failure injection when cluster starts...
- do the above but only to some tests (like the replace tests), the non-replace-specific tests here can keep using `server_stop_gracefully`.

Note I don't want to increase the scope of this PR so that I would also leave to a follow-up issue, which is probably low priority, but still it's an issue. Please create one after the PR is merged.

Reply:
> I don't like this.
> 
> We should be testing cluster behavior under abrupt failures (e.g. `kill -9`), we can't rely on graceful shutdowns to happen each time.
> 
> Each time I review this PR I have more and more of this feeling that we're doing something wrong...
> 
> First of all, using gossiper liveness information is of course unreliable. What happens when node is marked down by gossiper, not because the node is down, but is behind a network partition? What will the topology coordinator do in this case? I guess that it will attempt to do the replace operation and ban the node in the process. What if the node comes back in the middle of replace? Will it disrupt the cluster while replace is ongoing? Will it send/receive data path operations? Maybe the correct way to handle it is to ban it _at the beginning_ of replace operation, not at the end as it is now (if I remember correctly)? @gleb-cloudius WDYT?

Of course. This was always the final plan. We also need to ban all nodes in the ignored_node list. We just do not have a mechanism to do it since the one we have relies on the `left` state (IIRC).

Reply:
If something is in "the final plan", we should have an issue about it, not keep it in one or two people's heads. (It definitely skipped my awareness until now.) I don't think we have one. @patjed41 please do create an issue about this.

Reply:
> If something is in "the final plan", we should have an issue about it, not keep it in one or two people's heads. (It definitely skipped my awareness until now.) I don't think we have one. @patjed41 please do create an issue about this.

At least ignored_nodes isolation is in the failure recovery doc:

```
def topology_change():
   isolate_nodes(ignored_nodes)
   version++
```

The plan was to reuse the same `isolate_nodes` that we were going to use during a failure while recovering from a previous one, but in the end that was postponed/canceled, so we did not write `isolate_nodes`.

Reply:
Do we have the same plan for removenode? That is, banning the node being removed and the ignored nodes. If yes, I think we can create one issue for both replace and removenode.

If I remember correctly, we also want to reject removenode if the node being removed is alive. We already do this in `storage_service::raft_removenode`, but I think we should also (or only) do it on the topology coordinator's side since the gossip state might change. If I'm right, should I create an issue for this?

Reply:
> Do we have the same plan for removenode?

Of course.

> That is, banning the node being removed and the ignored nodes. If yes, I think we can create one issue for both replace and removenode.
> 
> If I remember correctly, we also want to reject removenode if the node being removed is alive. We already do this in `storage_service::raft_removenode`, but I think we should also (or only) do it on the topology coordinator's side since the gossip state might change. If I'm right, should I create an issue for this?

Yes.

Reply:
> Second, regarding the abrupt stop. We should have tests with abrupt stop. At least one. Of course waiting 20 seconds because of the user-experience gossiper-based protection is not an option. I see the following options:
> - this 20 seconds delay is failure detector setting, which can be configured. We could move these replace tests into `topology_custom` and start clusters there with custom configuration, reducing the delay. This should go together with disabling the tests in debug mode, where the low delay combined with latencies may cause failure detector "flickering"
> - OR disable the protection using a failure injection, but that probably also requires moving the tests to `topology_custom`, to enable the failure injection when cluster starts...
> - do the above but only to some tests (like the replace tests), the non-replace-specific tests here can keep using `server_stop_gracefully`.

I like the dot 1 + dot 3 approach the most. By "the replace tests" you probably mean `test_replace.py` and `test_raft_ignore_nodes.py`. That will leave us with only one `server_stop -> server_stop_gracefully` change but it's fine.

Reply:
I opened #16108 and #16109

Reply:
> I opened https://github.com/scylladb/scylladb/issues/16108 and https://github.com/scylladb/scylladb/issues/16109

Great, thanks.